README.md
# Introduction

[![travis](https://img.shields.io/travis/adrai/sessionstore.svg)](https://travis-ci.org/adrai/sessionstore) [![npm](https://img.shields.io/npm/v/sessionstore.svg)](https://npmjs.org/package/sessionstore)

Sessionstore is a node.js module for multiple databases. It can be very useful if you work with express or connect.

# Installation

    $ npm install sessionstore

# Usage

## Connecting to in-memory

	var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore()
    }));

## Connecting to mongodb

    var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore({
            type: 'mongodb',
            host: 'localhost',         // optional
            port: 27017,               // optional
            dbName: 'sessionDb',       // optional
            collectionName: 'sessions',// optional
            timeout: 10000             // optional
            // authSource: 'authedicationDatabase',        // optional
      	    // username: 'technicalDbUser',                // optional
      	    // password: 'secret'                          // optional
            // url: 'mongodb://user:pass@host:port/db?opts // optional
        })
    }));

## Connecting to tingodb

    var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore({
            type: 'tingodb',
            dbPath: __dirname + '/',   // optional
            collectionName: 'sessions',// optional
            timeout: 10000             // optional
        })
    }));

## Connecting to couchdb

    var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore({
            type: 'couchdb',
            host: 'http://localhost',  // optional
            port: 5984,                // optional
            dbName: 'express-sessions',// optional
            collectionName: 'sessions',// optional
            timeout: 10000             // optional
        })
    }));

## Connecting to redis

    var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore({
            type: 'redis',
            host: 'localhost',         // optional
            port: 6379,                // optional
            prefix: 'sess',            // optional
            timeout: 10000             // optional
        })
    }));

## Connecting to memcached

    var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore({
            type: 'memcached',
            host: 'localhost',         // optional
            port: 11211,               // optional
            prefix: 'sess',            // optional
            retries: 2,                // optional
            failover: false,           // optional
            failoverTime: 60,          // optional
            timeout: 10000             // optional
        })
    }));

## Connecting to elasticsearch (https://github.com/elastic/elasticsearch)

    var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore({
            type: '@elasticsearch',
            node: 'localhost:9200',    // optional
            prefix: '',                // optional
            index: 'express',          // optional
            typeName: 'session',       // optional
            pingInterval: 1000,        // optional
            timeout: 10000             // optional
        })
    }));


## Connecting to elasticsearch legacy (https://github.com/elastic/elasticsearch-js-legacy)

    var sessionstore = require('sessionstore');

    var express = require('express');
    var app = express();

    app.use(express.session({
        store: sessionstore.createSessionStore({
            type: 'elasticsearch',
            host: 'localhost:9200',    // optional
            prefix: '',                // optional
            index: 'express',          // optional
            typeName: 'session',       // optional
            pingInterval: 1000,        // optional
            timeout: 10000             // optional
        })
    }));

## Catch connect and disconnect events

    var ss = sessionstore.createSessionStore({ type: 'mongodb' }, function(err, ss) {
        console.log('hello from callback');
        // use store here...
        // app.use(express.session({
        //     store: ss
        // }));
    });
    ss.on('connect', function() {
        console.log('hello from event');
        // or here
        // app.use(express.session({
        //     store: ss
        // }));
    });
    ss.on('disconnect', function() {
        console.log('bye');
    });

#[Release notes](https://github.com/adrai/sessionstore/blob/master/releasenotes.md)

# Database Support
Currently these databases are supported:

1. inmemory
2. mongodb ([node-mongodb-native] (https://github.com/mongodb/node-mongodb-native))
3. couchdb ([cradle] (https://github.com/cloudhead/cradle))
4. tingodb ([tingodb] (https://github.com/sergeyksv/tingodb))
5. redis ([redis] (https://github.com/mranney/node_redis))
6. memcached ([memjs] (https://github.com/alevy/memjs))
7. elasticsearch ([elasticsearch] (https://github.com/elastic/elasticsearch-js))


# License

Copyright (c) 2019 Adriano Raiano

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


releasenotes.md
### [v1.3.7](https://github.com/adrai/sessionstore/compare/v1.3.6...v1.3.7)
- mognodb: default for useUnifiedTopology [#55](https://github.com/adrai/sessionstore/pull/55) thanks to [rohantarai](https://github.com/rohantarai)

### [v1.3.6](https://github.com/adrai/sessionstore/compare/v1.3.5...v1.3.6)
- mongodb: remove unwanted debug call

### [v1.3.5](https://github.com/adrai/sessionstore/compare/v1.3.4...v1.3.5)
- mongodb: add ignoreConcurrency flag

### [v1.3.4](https://github.com/adrai/sessionstore/compare/v1.3.3...v1.3.4)
- upgrade versions of lodash and async [#51](https://github.com/adrai/sessionstore/pull/51) thanks to [andrefbsantos](https://github.com/andrefbsantos)

### [v1.3.3](https://github.com/adrai/sessionstore/compare/v1.3.2...v1.3.3)
- remove ttl parameter in elasticsearch driver

### [v1.3.2](https://github.com/adrai/sessionstore/compare/v1.3.1...v1.3.2)
- update mocha and loadash libraries, regarding npm audit vulnerabilities

### [v1.3.1](https://github.com/adrai/sessionstore/compare/v1.3.0...v1.3.1)
- remove deprecated option max_attempts from redis options

### [v1.3.0](https://github.com/adrai/sessionstore/compare/v1.2.18...v1.3.0)
- compatibility with new mongodb version (3.x)

### [v1.2.18](https://github.com/adrai/sessionstore/compare/v1.2.16...v1.2.18)
- fix for new mongodb driver

### [v1.2.16](https://github.com/adrai/sessionstore/compare/v1.2.15...v1.2.16)
- recover elasticsearch connection after temp ping error [#39](https://github.com/adrai/sessionstore/issues/39) [#40](https://github.com/adrai/sessionstore/pull/40) thanks to [ewjmulder](https://github.com/ewjmulder)

### [v1.2.15](https://github.com/adrai/sessionstore/compare/v1.2.14...v1.2.15)
- redis, mongodb: call disconnect on ping error

### [v1.2.14](https://github.com/adrai/sessionstore/compare/v1.2.13...v1.2.14)
- Support mongo connection string

### [v1.2.13](https://github.com/adrai/sessionstore/compare/v1.2.12...v1.2.13)
- redis, mongodb: call disconnect on ping error

### [v1.2.12](https://github.com/adrai/sessionstore/compare/v1.2.11...v1.2.12)
- redis: added optional heartbeat

### [v1.2.11](https://github.com/adrai/sessionstore/compare/v1.2.9...v1.2.11)
- some updates

### [v1.2.9](https://github.com/adrai/sessionstore/compare/v1.2.8...v1.2.9)
- redis: fix for new redis lib

### [v1.2.8](https://github.com/adrai/sessionstore/compare/v1.2.7...v1.2.8)
- mongodb: added optional heartbeat

### [v1.2.7](https://github.com/adrai/sessionstore/compare/v1.2.6...v1.2.7)
- couchdb: fixed undefined is not function issue for expires [#28](https://github.com/adrai/sessionstore/issues/28) thanks to [priyanknarvekar](https://github.com/priyanknarvekar)

### [v1.2.6](https://github.com/adrai/sessionstore/compare/v1.2.5...v1.2.6)
- mongodb: give possibility to use authSource

### [v1.2.5](https://github.com/adrai/sessionstore/compare/v1.2.4...v1.2.5)
- redis: implemented touch

### [v1.2.4](https://github.com/adrai/sessionstore/compare/v1.2.3...v1.2.4)
- optimization for `npm link`'ed development

### [v1.2.3](https://github.com/adrai/sessionstore/compare/v1.1.0...v1.2.3)
- added elasticsearch support

### [v1.1.0](https://github.com/adrai/sessionstore/compare/v1.0.6...v1.1.0)
- session expiration comes from cookie now

### v1.0.6
- added mongodb driver 2.x support

### v1.0.5
- replace json-serialize with jsondate

### v1.0.4
- fix memcached implementation

### v1.0.3
- parse json with json-serialize


lib/databases/redis.js
var util = require('util'),
    Session = require('../sessionInterface'),
    use = require('../use'),
    _ = require('lodash'),
    async = require('async'),
    jsondate = require('jsondate');

var redis, Redis;
try {
  Redis = use('ioredis');
} catch (e) {
  try {
    redis = use('redis');
  } catch (e2) {
    throw e;
  }
}

var RedisSessionStore = function (options) {
  options = options || {};
  Session.Store.call(this, options);

  var defaults = {
    host: 'localhost',
    port: 6379,
    family: 4,
    prefix: 'sess',
    ttl:  60 * 60 * 24 * 14, // 14 days
    retry_strategy: function (options) {
      return undefined;
    }//,
    // heartbeat: 60 * 1000
  };

  _.defaults(options, defaults);

  if (options.url) {
    var url = require('url').parse(options.url);
    if (url.protocol === 'redis:') {
      if (url.auth) {
        var userparts = url.auth.split(':');
        options.user = userparts[0];
        if (userparts.length === 2) {
          options.password = userparts[1];
        }
      }
      options.host = url.hostname;
      options.port = url.port;
      if (url.pathname) {
        options.db = url.pathname.replace('/', '', 1);
      }
    }
  }

  if (Redis) {
    options.maxRetriesPerRequest = options.maxRetriesPerRequest || null;
    delete options.retry_strategy;
    options.reconnectOnError = options.reconnectOnError || function () { return false; };
    if (options.sentinels) {
      delete options.host;
      delete options.port;
    }
  }

  this.options = options;
}

util.inherits(RedisSessionStore, Session.Store);

_.extend(RedisSessionStore.prototype, {

  connect: function (callback) {
    var self = this;

    var options = this.options;

    if (Redis) {
      this.client = new Redis(_.omit(options, 'prefix'));
    } else {
      this.client = new redis.createClient(options.port || options.socket, options.host, _.omit(options, 'prefix'));
    }

    this.prefix = options.prefix;

    var calledBack = false;

    if (!Redis) {
      if (options.password) {
        this.client.auth(options.password, function(err) {
          if (err && !calledBack && callback) {
            calledBack = true;
            if (callback) callback(err, self);
            return;
          }
          if (err) throw err;
        });
      }

      if (options.db) {
        this.client.select(options.db);
      }
    }

    this.client.on('end', function () {
      self.disconnect();
      self.stopHeartbeat();
    });

    if (Redis) {
      this.client.on('close', function () {
        self.disconnect();
        self.stopHeartbeat();
      });
    }

    this.client.on('error', function (err) {
      console.log(err);

      if (calledBack) return;
      calledBack = true;
      if (callback) callback(null, self);
    });

    this.client.on('connect', function () {
      if (options.db && !Redis) {
        self.client.send_anyways = true;
        self.client.select(options.db);
        self.client.send_anyways = false;
      }

      self.emit('connect');

      if (self.options.heartbeat) {
        self.startHeartbeat();
      }

      if (calledBack) return;
      calledBack = true;
      if (callback) callback(null, self);
    });
  },

  stopHeartbeat: function () {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
      delete this.heartbeatInterval;
    }
  },

  startHeartbeat: function () {
    var self = this;

    var gracePeriod = Math.round(this.options.heartbeat / 2);
    this.heartbeatInterval = setInterval(function () {
      var graceTimer = setTimeout(function () {
        if (self.heartbeatInterval) {
          console.error((new Error ('Heartbeat timeouted after ' + gracePeriod + 'ms (redis)')).stack);
          self.disconnect();
        }
      }, gracePeriod);

      self.client.ping(function (err) {
        if (graceTimer) clearTimeout(graceTimer);
        if (err) {
          console.error(err.stack || err);
          self.disconnect();
        }
      });
    }, this.options.heartbeat);
  },

  disconnect: function (callback) {
    this.stopHeartbeat();

    if (this.client) {
      this.client.end(true);
    }
    this.emit('disconnect');
    if (callback) callback(null, this);
  },

  set: function (sid, sess, callback) {
    var prefixedSid = this.prefix + ':' + sid;

    try {
      var ttl = this.options.ttl;
      if (sess && sess.cookie && sess.cookie.expires) {
        var expInMs = (new Date(sess.cookie.expires)).getTime() - Date.now();
        if (expInMs < 1000) {
          expInMs = 1000;
        }
        ttl = Math.round(expInMs / 1000);
      }
      sess = JSON.stringify(sess);
      this.client.setex(prefixedSid, ttl, sess, callback  || function () {});
    } catch (err) {
      if (callback) callback(err);
    }
  },

  touch: function (sid, sess, callback) {
    var prefixedSid = this.prefix + ':' + sid;
    var ttl = this.options.ttl;
    if (sess && sess.cookie && sess.cookie.expires) {
      var expInMs = (new Date(sess.cookie.expires)).getTime() - Date.now();
      if (expInMs < 1000) {
        expInMs = 1000;
      }
      ttl = Math.round(expInMs / 1000);
    }

    this.client.expire(prefixedSid, ttl, callback  || function () {});
  },

  get: function (sid, callback) {
    var prefixedSid = this.prefix + ':' + sid;

    this.client.get(prefixedSid, function (err, data) {
      if (err) {
        if (callback) callback(err);
        return;
      }
      if (!data) {
        if (callback) callback(null, null);
        return;
      }

      var result;

      try {
        result = jsondate.parse(data.toString());
      } catch (error) {
        if (callback) callback(err);
        return;
      }

      if (callback) callback(null, result);
    });
  },

  destroy: function (sid, callback) {
    var prefixedSid = this.prefix + ':' + sid;
    this.client.del(prefixedSid, callback || function () {});
  },

  all: function (callback) {
    var self = this;
    this.client.keys(this.prefix + ':*', function(err, docs) {
      async.map(docs, function(doc, callback) {
        self.client.get(doc, function (err, data) {
          if (err) {
            if (callback) callback(err);
            return;
          }
          if (!data) {
            if (callback) callback(null, null);
            return;
          }

          var result;

          try {
            result = jsondate.parse(data.toString());
          } catch (error) {
            if (callback) callback(err);
            return;
          }

          if (callback) callback(null, result);
        });
      }, callback);
    });
  },

  length: function (callback) {
    this.client.keys(this.prefix + ':*', function (err, docs) {
      if (err) {
        if (callback) callback(err);
        return;
      }

      if (callback) callback(null, docs.length);
    });
  },

  clear: function (callback) {
    var self = this;
    this.client.keys(this.prefix + ':*', function(err, docs) {
      async.each(docs, function(doc, callback) {
        self.client.del(doc, callback);
      }, callback || function () {});
    });
  }

});

module.exports = RedisSessionStore;
